### DISCUSSION TITLE
[Security] Zen as a default browser + Workspaces

### DISCUSSION THREAD

**Discussion by Mikopet** [zen-browser/desktop#2422] [Ideas]
### Captchas

- [X] I have read the instructions.
- [X] I have searched existing issues and avoided creating duplicates.
- [X] I am not filing an enhancement request.

### What happened?

Hey!

I sort of ported myself almost entirely to Zen. Which I am happy about, even if it's not really a recommended way yet.

Anyhow, while I'm aware other browsers usually don't even have these options (I'm pointing to the containers, but includes workspaces too as they have a correlation), I'd consider it a nice baseline for modern web browsers.

I'm also aware that workspaces just experimental feature. However, I noticed a few interesting way of working around them, what should not work this way in my opinion. (or at least should be configurable from settings)

_(This issue will be a bit hectic, jumping between features and general speaking about security. Sorry for that.
Also these findings are opinions by me, not some wide accepted security best practices, handle as such. This is just the way I see these as a fellow browser developer in a different ecosystem.)_

---

#### 1a) Opening pages from external sources while browser is closed
- **ACTION:** The browser is closed, but will be opened by the system when clicking on a link somewhere.
- **RESULT:** In my first test it opened in the last used(/active) workspace. After upgrading to latest (`1.0.1-a.10`) it seemed to me, it opens in the default workspace. But the third test showed it was just some random event, I cannot reproduce again. It opens always in the last active space.
- **SUGGESTION:** external sources should open in the default workspace by default setting, and should be overridable in the settings menu.

_p.s.: it seems, it opens in the default container, what feels sort of right, if the default container is not particularly used for sessions. From security perspective these pages should be isolated._

#### 1b) Opening pages from external sources while browser is open
- **ACTION:** The browser is open, user clicks on a link somewhere outside of it.
- **RESULT:** same as `Case 1a`. Opens in the active workspace. The only difference I see here, is it opens in the container of the workspace.
- **SUGGESTION:** Same as `Case 1a`. external sources should open in the default workspace by default setting, and should be overridable in the settings menu.

_p.s.: although this is the same setting, behavior is slightly different. The issue is also not entirely the same. I'm coming back to this in `Case 2-3`._

---

### Case 2: Workspace container enforcing
This is where things get tricky. So, we know containers and workspaces are not the same. But they sort of should be able to behave that way. Perhaps. Perhaps not.

I guess it depends on preference, how we have mental models about the tabs and sessions, cookies, logins, cross-site thingies. And this is a rabbit-hole, where complexity suddenly can increase.

#### 2a) New tab opens in a workspace externally
- **ACTION:** The browser is closed, but will be opened by the system when clicking on a link somewhere. (`Case 1a`)
- **RESULT:** same as `Case 1a`. Opens in the active workspace. With no container.
- **SUGGESTION:** Default working is to open new tabs in the default container of the workspace.

_p.s.: this is a bug in my opinion. but also points out a conceptual question, what is mentioned in `Case 1a`._

#### 2b) New tab opens in a workspace internally
- **ACTION:** A tab was moved from other workspace.
- **RESULT:** The tab moves to the new designation, but the container stays the same.
- **SUGGESTIONS:**
  - Default working could be this, no issue. **It makes some sense not to change the container setting arbitrarily.**
  - However, if the tab was in no container, it should be put into a container. (inherited from the workspace)
  - These ways of working also should be a user setting. Like I want every tab to be put in the associated container of the workspace.
 
_p.s.: this setting is more for mental model, than technical security. But the false sense of security knowing that workspace has default container makes user to easily overlook container settings during sessions. Also I see there was something in the last release "Disabled forcing container grouping for workspaces". I'm not sure it is related or not._

---

### Case 3: Containers and their storages
If the previous case looked tricky, wait until read this. 
- **ACTION:** Both no container and in one container user is logged in to a website. User decides to clear cookies and site data.
- **RESULT:** User is logged out in all of the containers.

This is kind of expected behavior from the technical perspective. But not intuitive. Allegedly these containers are isolated, but then this working doesn't feel like that. I don't know how Mozilla stores these "isolated" sessions under the hood, but this is conflicting mental model.

In addition, I cannot clear all cookies and site data in one particular container.

- **SUGGESTIONS:**
  - There should be two different button for clearing cookies and site data only in that particular container, or globally the browser.
  - In the settings around the clear and manage data and exceptions, there should be the containers integrated. For instance, if user wants to clean only just the personal or only just the work container. Makes sense.
  - There should be an option to NOT having no-container mode. (with that said that is also a container, isn't it?)
  - Also containers should be able to set up to behave differently. Like clean all data after sessions. It is similar to the second suggestion, but related to global browser settings too. (like clean history and whatnot, but only in selected containers. I know, this could be extremely tricky, if the history is not isolated. I did not check this.)

_p.s.: There is a checkbox in the container settings: "Select a container for each new tab". I'm not sure what it meant to do, but for me it seems it is broken, and has no effect. At least not in the aforementioned cases. However, the point is here I believe is again, the mental model. You have to put that website into one bucket. If it not fits, then open it in a random bucket._

---

### Summary

As you see these issues are correlated, and sort of complex if somebody wants to wrap their head around it. But the suggestions I made (especially about the settings) makes a lot of sense to me. 

Usually people have workspaces like personal/work/etc. This kind of covers the default containers. I also like having a random workspace what is the default WS, and has no default container. This is mostly for ephemeral things. Sure, it should better have a container for such cases, because no-container tabs could do unexpected behavior based on some settings mentioned above.

But the way I see, a well defined and secure browsing experience incorporates a good isolation, where multiple personal workspaces could be put under a `Personal` container, multiple workspaces for job/work usage could be put under one/more `Work` containers (depending on the usecase ofc), and so on. (Sure, there are also profiles, what gives better isolation. For sure. But let's not delve into that topic right now.)

I also think the same good and secure practice looks like when user is opening a website from external source, it opens in the default workspace (or based on user setting), what is by concept sort of permissionless. The baseline is not having any cookies or site data attached to it.

Then when the user moves that tab into a workspace where mentally they want it to exists, then workspace forces the container to the tab. Independent of it was already in a container or not. (or based on the user setting)

Which leads to the end, where you want to clean up the residuals of these ephemeral site data, so clean only data only in containers, or only in workspaces (based on user settings, I guess)

I hope this "short" review is something useful for you. (it took me a considerable amount of time!) I wish I could help with the code, but I don't know yet where to look.

Anyways, amazing browser. Kudos! :tada: 

### Reproducible?

- [ ] I have checked that this issue cannot be reproduced on Mozilla Firefox.

### Version

1.0.1-a.10

### What platform are you seeing the problem on?

Linux

**Comment by linear** [comment]
<p><a href="https://linear.app/zen-browser/issue/ZEN-2029/[security]-zen-as-a-default-browser-workspaces">ZEN-2029 [Security] Zen as a default browser + Workspaces</a></p>

**Comment by RayZ3R0** [comment]
This is more of a suggestion than a bug. Moving this to discussions
